### PR TITLE
prov/rxm: Use separate list of the posted RX buffers per connection

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
This helps to avoid leak of the RX buffers that were preposted for closed
MSG EP in case of bi-directional connection (passive side of connection).

Before the connection is established and new MSG EP is allocated, move
old preposted RX buffers to the temporarily list and then cleanup the list.

When we have multiple spawned and then destroyed processes, the memory growths for every spawned processes. 
The following situation:
`spawn 1 proc -> destroy 1 proc -> spawn 2 proc -> destroy 2 proc -> ... -> spawn N proc -> destroy N proc
`would lead that (N * (rx-attr->size)) RX buffers will be available in RX buffers pool.
In this situation we can re-use allocated buffers that are no longer needed for the destroyed connection.
With this PR, RxM will allocate only (1 * (rx_attr->size)) RX buffers, if the sequence is followed as it in my example above.

Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>

### CODE REVIEW

File: prov/rxm/src/rxm_conn.c [hunk 1]
```diff
@@ -82,7 +82,8 @@ static int rxm_msg_ep_open(struct rxm_ep *rxm_ep, struct fi_info *msg_info,
 	}
 
 	if (!rxm_ep->srx_ctx) {
-		ret = rxm_ep_prepost_buf(rxm_ep, msg_ep);
+		dlist_init(&rxm_conn->posted_rx_list);
```
Comment (shefty) [inline prov/rxm/src/rxm_conn.c]: Please move list initialization to rxm_conn creation, to avoid potential issues down the road.

File: prov/rxm/src/rxm_conn.c [hunk 2]
```diff
@@ -97,9 +98,13 @@ static int rxm_msg_ep_open(struct rxm_ep *rxm_ep, struct fi_info *msg_info,
 static void rxm_conn_close(struct util_cmap_handle *handle)
 {
 	struct rxm_conn *rxm_conn = container_of(handle, struct rxm_conn, handle);
+
 	if (!rxm_conn->msg_ep)
 		return;
-
+	dlist_init(&rxm_conn->saved_posted_rx_list);
```
Comment (shefty) [inline prov/rxm/src/rxm_conn.c]: same here -- move to rxm_conn creation.

File: prov/rxm/src/rxm.h [hunk 3]
```diff
@@ -115,9 +115,11 @@ struct rxm_fabric {
 struct rxm_conn {
 	struct fid_ep *msg_ep;
 	struct dlist_entry postponed_tx_list;
+	struct dlist_entry posted_rx_list;
 	struct util_cmap_handle handle;
 	/* This is saved MSG EP fid, that hasn't been closed during
 	 * handling of CONN_RECV in CMAP_CONNREQ_SENT for passive side */
+	struct dlist_entry saved_posted_rx_list;
```
Comment (shefty) [inline prov/rxm/src/rxm.h]: comment refers to msg ep, not rx_list.  Please move declaration down,

File: prov/rxm/src/rxm.h [hunk 4]
```diff
@@ -396,7 +398,7 @@ struct rxm_ep {
 
 	struct rxm_buf_pool	buf_pools[RXM_BUF_POOL_MAX];
 
-	struct dlist_entry	post_rx_list;
+	struct dlist_entry	srx_posted_rx_list;
```
Comment (shefty) [inline prov/rxm/src/rxm.h]: nit: this could just be named posted_srx_list.  Can the name change go into a separate commit?

**Comment by a-ilango** [comment]
Is the patch trying to fix the assertion failure when closing util buf pool? I wonder if we should add a flag to skip that check. As long as we don't do any additional processing when releasing a buffer back to the pool, it should be safe to remove the assertion. This would enable us to avoid additional book-keeping for the buffers.

**Comment by gladkovdmitry17** [comment]
> Is the patch trying to fix the assertion failure when closing util buf pool?

@a-ilango, No, I didn't see this issue. All buffers were returned into their place, because when the RxM EP are closing, it releases all allocated RX buffers. 

This patch fixes the following problem:
There are MPI application that uses master<->slave approach by spawning new MPI processes.
Whens spawned MPI process is being killed, the master and other slaves still has allocated RX buffers that are related to this connection.
The solution proposed in the patch is to extend the handling of the `FI_SHUTDOWN` event by releasing all allocated RX buffers related to the connection.

**Comment by a-ilango** [comment]
> The solution proposed in the patch is to extend the handling of the FI_SHUTDOWN event by releasing all allocated RX buffers related to the connection.

The only reason we release all allocated buffers because util_buf_pool would complain about buffers being used when we don't release it, right? There won't be any memory leak in any case as util buf pool frees all allocated buffers.

**Comment by gladkovdmitry17** [comment]
> The only reason we release all allocated buffers because util_buf_pool would complain about buffers being used when we don't release it, right?

Yes, it's right. But RxM keep the RX buffers allocated for the connection the even if the connection was destroyed.

**Comment by a-ilango** [comment]
> But RxM keep the RX buffers allocated for the connection the even if the connection was destroyed

So is it more of a memory usage issue? If that's not a problem, we can skip releasing them and update util buf pool to ignore unreleased buffers or if these changes are not in fast path or have minimal impact, then we can merge them.

**Comment by gladkovdmitry17** [comment]
> If that's not a problem, we can skip releasing them and update util buf pool to ignore unreleased buffers or if these changes are not in fast path or have minimal impact, then we can merge them.

Yes, we can skip releasing RX buffers until the util buffer pool is destroyed (current behavior). Since the util buffer pool growth when there is no available space, but it doesn't free buffers (i.e. don't reduce memory consumption), when this memory is no longer needed.  This is OK.
But when we have multiple spawned and then destroyed processes, the memory growths for every spawned processes. I mean the following situation:
spawn 1 proc -> destroy 1 proc -> spawn 2 proc -> destroy 2 proc -> ... -> spawn N proc -> destroy N proc.
It would lead that (N * (rx-attr->size)) RX buffers will be available in RX buffers pool.
But in this situation we can re-use allocated buffers that are no longer needed to destroyed connection.
With this PR, RxM will allocate only (1 * (rx_attr->size)) RX buffers, if the sequence is followed as it in my example above.

**Comment by a-ilango** [comment]
Got it. That kind of scenario would result in high memory usage. Thanks for the explanation!

**Comment by gladkovdmitry17** [comment]
> That kind of scenario

I have noticed this in the MPICH/spawn/task_master test, when was analyzing RxM behavior

**Comment by shefty** [comment]
Having that level of detail in the commit message would have been helpful in understanding the reasoning for the change and the impact.

**Comment by gladkovdmitry17** [comment]
> Having that level of detail in the commit message would have been helpful in understanding the reasoning for the change and the impact.

I've updated the PR details to capture the main point that is addressed in the PR.
